File: typescript/src/p1056-confusing-number/solution.test.ts
import { describe, test, expect } from "vitest";
import { solutions } from "./solution";

describe.each(solutions)("confusingNumber - %o", (confusingNumber) => {
  // --- Provided examples ---

  test("example 1: 6 is confusing (rotates to 9)", () => {
    expect(confusingNumber(6)).toBe(true);
  });

  test("example 2: 89 is confusing (rotates to 68)", () => {
    expect(confusingNumber(89)).toBe(true);
  });

  test("example 3: 11 is not confusing (rotates to 11)", () => {
    expect(confusingNumber(11)).toBe(false);
  });

  test("example 4: 25 is not confusing (contains invalid digit)", () => {
    expect(confusingNumber(25)).toBe(false);
  });

  // --- Edge cases ---

  test("0 rotated is still 0", () => {
    expect(confusingNumber(0)).toBe(false);
  });

  test("1 rotated is still 1", () => {
    expect(confusingNumber(1)).toBe(false);
  });

  test("8 rotated is still 8", () => {
    expect(confusingNumber(8)).toBe(false);
  });

  // --- Invalid digits ---

  test("2 is not confusing (invalid digit)", () => {
    expect(confusingNumber(2)).toBe(false);
  });

  test("3 is not confusing (invalid digit)", () => {
    expect(confusingNumber(3)).toBe(false);
  });

  test("4 is not confusing (invalid digit)", () => {
    expect(confusingNumber(4)).toBe(false);
  });

  test("5 is not confusing (invalid digit)", () => {
    expect(confusingNumber(5)).toBe(false);
  });

  test("7 is not confusing (invalid digit)", () => {
    expect(confusingNumber(7)).toBe(false);
  });

  // --- Valid digits but same after rotation ---

  test("69 rotated is 69 (not confusing)", () => {
    expect(confusingNumber(69)).toBe(false);
  });

  test("96 rotated is 96 (not confusing)", () => {
    expect(confusingNumber(96)).toBe(false);
  });

  test("88 rotated is 88 (not confusing)", () => {
    expect(confusingNumber(88)).toBe(false);
  });

  // --- Confusing multi-digit numbers ---

  test("10 is confusing (rotates to 01 = 1)", () => {
    expect(confusingNumber(10)).toBe(true);
  });

  test("60 is confusing (rotates to 09 = 9)", () => {
    expect(confusingNumber(60)).toBe(true);
  });

  test("900 is confusing (rotates to 006 = 6)", () => {
    expect(confusingNumber(900)).toBe(true);
  });

  test("689 rotated is 689 (not confusing)", () => {
    expect(confusingNumber(689)).toBe(false);
  });

  test("680 is confusing (rotates to 089 = 89)", () => {
    expect(confusingNumber(680)).toBe(true);
  });

  // --- Larger numbers ---

  test("1000000006 is confusing", () => {
    expect(confusingNumber(1000000006)).toBe(true);
  });

  test("123 is not confusing (contains invalid digits)", () => {
    expect(confusingNumber(123)).toBe(false);
  });
});

File: problems/p1056_confusing_number/java/p1056_confusing_number/_SolutionTest.java
package p1056_confusing_number;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.*;

class _SolutionTest {

    static Stream<Solution> solutions() {
        return Stream.of(
                new Default()
        );
    }

    // --- Provided examples ---

    @ParameterizedTest
    @MethodSource("solutions")
    void example1_6isConfusing(Solution solution) {
        assertTrue(solution.confusingNumber(6));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void example2_89isConfusing(Solution solution) {
        assertTrue(solution.confusingNumber(89));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void example3_11isNotConfusing(Solution solution) {
        assertFalse(solution.confusingNumber(11));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void example4_25isNotConfusing(Solution solution) {
        assertFalse(solution.confusingNumber(25));
    }

    // --- Edge cases ---

    @ParameterizedTest
    @MethodSource("solutions")
    void zero_rotatedIsSame(Solution solution) {
        assertFalse(solution.confusingNumber(0));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void one_rotatedIsSame(Solution solution) {
        assertFalse(solution.confusingNumber(1));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void eight_rotatedIsSame(Solution solution) {
        assertFalse(solution.confusingNumber(8));
    }

    // --- Invalid digits ---

    @ParameterizedTest
    @MethodSource("solutions")
    void containsTwo_notConfusing(Solution solution) {
        assertFalse(solution.confusingNumber(2));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void containsThree_notConfusing(Solution solution) {
        assertFalse(solution.confusingNumber(3));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void containsFour_notConfusing(Solution solution) {
        assertFalse(solution.confusingNumber(4));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void containsFive_notConfusing(Solution solution) {
        assertFalse(solution.confusingNumber(5));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void containsSeven_notConfusing(Solution solution) {
        assertFalse(solution.confusingNumber(7));
    }

    // --- Valid digits but same after rotation ---

    @ParameterizedTest
    @MethodSource("solutions")
    void sixNine_rotatedIsSame(Solution solution) {
        assertFalse(solution.confusingNumber(69));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void nineSix_rotatedIsSame(Solution solution) {
        assertFalse(solution.confusingNumber(96));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void eightyEight_rotatedIsSame(Solution solution) {
        assertFalse(solution.confusingNumber(88));
    }

    // --- Confusing multi-digit numbers ---

    @ParameterizedTest
    @MethodSource("solutions")
    void ten_isConfusing(Solution solution) {
        assertTrue(solution.confusingNumber(10));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void sixZero_isConfusing(Solution solution) {
        assertTrue(solution.confusingNumber(60));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void nineHundred_isConfusing(Solution solution) {
        assertTrue(solution.confusingNumber(900));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void sixEightNine_rotatedIsSame(Solution solution) {
        assertFalse(solution.confusingNumber(689));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void sixEightZero_isConfusing(Solution solution) {
        assertTrue(solution.confusingNumber(680));
    }

    // --- Larger numbers ---

    @ParameterizedTest
    @MethodSource("solutions")
    void largeConfusingNumber(Solution solution) {
        assertTrue(solution.confusingNumber(1000000006));
    }

    @ParameterizedTest
    @MethodSource("solutions")
    void largeNonConfusingWithInvalidDigit(Solution solution) {
        assertFalse(solution.confusingNumber(123));
    }
}
